# <span style="color:red;"><center>TP - Les algorithmes gloutons </center></span>

## 1. Objectif du TP

Implémenter les algorithmes glouton vu en cours dans le cas des problèmes du rendu de monnaie et du sac à dos.

## 2. Le problème du rendu de monnaie

On utilise le **système de monnaie de la zone euro** pour laquelle nous disposons des **pièces de valeurs suivantes** (en centimes) :  
`S=[1, 2, 5, 10, 20, 50, 100, 200, 500]`

Ecrire une fonction `rendu_glouton(S,r)` qui **prend en argument une liste `S` des valeurs possibles pour les pièces et la somme `r` à rendre**. Le **résultat** est renvoyé sour la forme d'une **liste de la même taille que `S` dont les éléments indiquent le nombre de pièces à rendre de chaque sorte**.  
Par exemple, avec le système de monnaie de la zone euro et pour `r = 8`, la fonction renverra `[1, 1, 1, 0, 0, 0, 0, 0, 0]`. Ce qui veut dire que pour rendre 8 centimes, on rend une pièce de 1, une pièce de 2 et 1 pièce de 5 centimes.

In [19]:
def rendu_glouton(S,r) :
    """Prend en argument une liste S des valeurs possibles pour les pièces et la somme r à rendre.
    Le résultat est renvoyé sour la forme d'une liste de la même taille que S dont les éléments
    indiquent le nombre de pièces à rendre de chaque sorte."""
    n=len(S) # nombre de pièce différentes
    rendu = [0 for i in range(n)] #crée une liste de même taille que S mais avec des 0, pour le rendu
    reste=r #reste est la somme qui reste à rendre
    i = n-1 # i est l'indice de la pièce qu'on essaie de rendre, on commence par la plus grosse pièce
    while reste != 0 : #on continue tant qu'il reste quelque chose à rendre
        if S[i] <= reste : #si la valeur de la pièce est plus petite que ce qu'il reste à rendre
            rendu[i]+=1 #on ajoute 1 pièce de cette valeur au résultat
            reste=reste-S[i] #on met à jour la somme restant à rendre
        else : #sinon, on essaie la pièce juste en dessous
            i-=1
    return rendu #on renvoie le résultat
    

### Test 1
Tester votre algorithme en supposant que vous devez rendre la monnaie sur 5 euros pour un achat de 2,37€. La **somme à rendre est donc de 2,63€**. Le résultat obtenu vous semble-il optimum ? (il devrait !)

In [20]:
#Test 1
S=[1, 2, 5, 10, 20, 50, 100, 200, 500]
r=8
print("Liste des pièces à rendre pour la somme de",r,"centimes :",rendu_glouton(S,r))
r=263
print("Liste des pièces à rendre pour la somme de",r,"centimes :",rendu_glouton(S,r))

Liste des pièces à rendre pour la somme de 8 centimes : [1, 1, 1, 0, 0, 0, 0, 0, 0]
Liste des pièces à rendre pour la somme de 263 centimes : [1, 1, 0, 1, 0, 1, 0, 1, 0]


Pour rendre **8 centimes**, l'algorithme glouton indique donc de rendre $1$ pièce de *5 centimes*, $1$ pièce de *2 centimes* et $1$ pièce de *1 centime*, comme prévu. Soit **3 pièces**, ce qui est la **solution optimum**.  
Pour rendre **2€63**, l'algorithme glouton indique donc de rendre $1$ pièce de *2 euros*, $1$ pièce de *50 centimes*, $1$ pièce de *10 centimes*, $1$ pièce de *2 centimes* et $1$ pièce de *1 centime*. Soit **5 pièces**, ce qui est aussi une **solution optimum**, comme il se doit avec le système de monnaie de l'euro.

### Test 2

Imaginons à présent que nous sommes en Angleterre, avant 1971. Le système de monnaie, pour les pences, est `Sp=[1, 3, 4, 6, 12]`.

Tester votre algorithme pour **rendre la somme de 8 pences**. Le résultat est-il optimum ? (il ne devrait pas cette fois !)

In [21]:
#Test 2
Sp=[1, 3, 4, 6, 12]
r=8
print("Liste des pièces à rendre pour la somme de",r,"centimes :",rendu_glouton(Sp,r))

Liste des pièces à rendre pour la somme de 8 centimes : [2, 0, 0, 1, 0]


Dans ce cas, pour rendre **8 pences**, l'algorithme glouton indique donc de rendre $1$ pièce de *6 pences* et $2$ pièce de *1 pence*. Soit **3 pièces**, ce qui n'est **pas la solution optimum**.  
En effet la **solution optimum** serait ici de rendre $2$ pièces de *4 pences*.

## 3. Le problème du sac à dos

Ecrire une fonction `sacados_glouton(objets, poids, choix)` qui prend **en argument la liste des objets disponibles** `objets`, le **poids maximum** du sac à dos `poids` et le **choix de la stratégie** à appliquer.  
- `objets` est une **liste de listes** (une par objet) où **chaque objet est représenté par une liste** de la forme `["nom de l'objet", valeur de l'objet, poids de l'objet]`.
- `choix` vaut `1` si on choisit les objets en fonction de leur **valeur**, `2` si on les choisit en fonction de leur **poids** et `3` si on les choisit en fonction de leur **rentabilité** (rapport valeur sur poids).

La fonction **renvoie une liste** contenant les **noms des objets choisis** et la **valeur totale** du sac à dos ainsi rempli.

In [33]:
def sacados_glouton(objets,poids,choix) :
    """Prend en argument la liste des objets disponibles objets, le poids maximum du sac à dos poids et le choix de la stratégie à appliquer.
    objets est une liste de listes (une par objet) où chaque objet est représenté par une liste de la forme ["nom de l'objet", valeur de l'objet, poids de l'objet].
    choix vaut 1 si on choisit les objets en fonction de leur valeur, 2 si on les choisit en fonction de leur poids et 3 si on les choisit en fonction de leur rentabilité.
    La fonction renvoie une liste contenant les noms des objets choisis et la valeur totale du sac à dos ainsi rempli."""
    from operator import itemgetter #la fonction itemgetter permet de récupérer la valeur à trier pour chaque objet
    from copy import deepcopy #on a besoin de deepcopy pour faire une vraie copie de liste de listes
    n=len(objets) #n est le nombre d'objets
    objets_copie=deepcopy(objets) #on crée une copie pour ne pas modifier la liste d'objets
    #on commence par ajouter une colonne pour la rentabilité
    for i in range(n):
        objets_copie[i].append(objets_copie[i][1]/objets_copie[i][2]) #calcul de la rentabilité
    #Il faut trier en sens décroissant, sauf si le choix se fait sur le poids (on commence alors par les plus légers)
    decroissant = choix != 2 # si choix = 2 alors decroissant = False, sinon, il est vrai
    objets_tries=sorted(objets_copie, key=itemgetter(choix),reverse=decroissant) #tri la liste par ordre décroissant selon la colonne choix
    poids_sac = 0 # le poids du sac à dos
    i=0 #l'indice de l'objet examiné
    contenu_sac=[] #le contenu du sac est vide pour commencer
    valeur_sac=0 #la valeur des objets contenus dans le sac
    while poids_sac <= poids and i<n : #on remplit le sac tant que le poids maximal n'est pas dépassé et qu'il reste des objets à examiner
        if poids_sac + objets_tries[i][2] <= poids : #on teste si l'objet examiné peut entrer dans le sac à dos
            contenu_sac.append(objets_tries[i][0]) #on ajoute le nom de l'objet au contenu
            poids_sac=poids_sac+ objets_tries[i][2] #on ajoute le poids de l'objet choisi au poids du sac
            valeur_sac=valeur_sac+objets_tries[i][1]
        i+=1 #on passe à l'objet suivant
    return contenu_sac, valeur_sac #on renvoie le contenu du sac et sa valeur


### Test 3
On considère la **liste d'objets** suivante :

|Objet|Valeur (€)|Poids (kg)|Rentabilité (€/kg)|
|:-:|:-:|:-:|:-:|
|A|126|14|9|
|B|32|2|16|
|C|20|5|4|
|D|5|1|5|
|E|18|6|3|
|F|80|8|10|

Tester votre algorithme pour un **poids maximum de 15 kg** et avec les **3 choix possibles**. Quel est le **meilleur choix** dans ce cas là ?  
Sachant que la solution optimale permet ici d'emporter un **butin de 132€**, notre algorithme **glouton est-il optimum** ?

In [34]:
#Test 3
objets=[['A',126,14],['B',32,2],['C',20,5],['D',5,1],['E',18,6],['F',80,8]]
poids=15
choix=1
resultat=sacados_glouton(objets,poids,choix)
print("Les objets emportés sont",resultat[0],"pour une valeur de",resultat[1],"euros.")
choix=2
resultat=sacados_glouton(objets,poids,choix)
print("Les objets emportés sont",resultat[0],"pour une valeur de",resultat[1],"euros.")
choix=3
resultat=sacados_glouton(objets,poids,choix)
print("Les objets emportés sont",resultat[0],"pour une valeur de",resultat[1],"euros.")

Les objets emportés sont ['A', 'D'] pour une valeur de 131 euros.
Les objets emportés sont ['D', 'B', 'C', 'E'] pour une valeur de 75 euros.
Les objets emportés sont ['B', 'F', 'D'] pour une valeur de 117 euros.


On voit que la **solution trouvée n'est jamais optimum**, quelque soit le critère choisi (poids, valeur, rentabilité). Cependant, on s'en **approche de très près** avec le critère de la valeur (131€ au lieu de 132€).

La **solution optimum** consiste à choisir les objets **B, C et F** pour un poids de 15kg et une valeur de **132€**. On peut le vérifier en testant toutes les possibilités (force brute).